Implement .at property for Koalas DataFrames and Series

floscha opened the pull request:
As requested in #382, this PR implements pandas' `.at` property for Koalas DataFrames and Series.

ueshin commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
+
+        column = key[1] if len(key) > 1 else self._ks.name
+        if column is not None and column not in self._kdf._metadata.data_columns:
+            raise KeyError("'%s" % column)
```
just out of curiosity, what's `'` for?

floscha replied:
This is just to be consistent with the way pandas uses Python's `KeyError`, putting the missing key between `'`s. However, I've just realized I've missed the closing `'` and the `'`s are not even necessary to start with 🙈 So I've fixed this with 2dc8fa5.

ueshin commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
+               .where(F.col(self._kdf._metadata.index_columns[0]) == row)
+               .select(column))
+        if sdf.count() < 1:
+            raise KeyError("'%s" % row)
```
ditto.

floscha replied:
See comment above.

ueshin commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
+        if sdf.count() < 1:
+            raise KeyError("'%s" % row)
+
+        values = DataFrame(sdf).to_pandas().iloc[:, 0].values
```
Actually we don't need to create Koalas `DataFrame` here. Spark DataFrame has `toPandas()`.

floscha replied:
You're right. Fixed this with 2dc8fa5.

floscha commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
+        if self._ks is None and len(key) != 2:
+            raise TypeError("Use DataFrame.at like .at[row_index, column_name]")
+        if self._ks is not None and len(key) != 1:
+            raise TypeError("Use Series.at like .at[column_name]")
```
True. Fixed with 8137630.

floscha commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
+    def __getitem__(self, key):
+        from databricks.koalas.frame import DataFrame
+
+        if self._ks is None and len(key) != 2:
```
Sure, added with 09df062.

ueshin replied:
Could you add tests for such a case?

floscha replied:
Sure, done with fb3587d.

floscha commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
+
+        row = key[0]
+        sdf = (series._kdf._sdf
+               .where(F.col(self._kdf._metadata.index_columns[0]) == row)
```
I've added a check now that raises an exception if the index level doesn't equal 1 with a61f163. I'll have to look closer into how `.at` behaves for multilevel indices which generally seems to be rather uncommon. Also, is a Koalas DataFrame without an index even valid? Like, wouldn't it for example crash when calling `to_pandas()` on it?

ueshin replied:
As for multi- or no index, currently `.loc` raises an exception for multi- or no index. Maybe we can follow the behavior for now, but please make sure to check it and raise the exception.
And yes, no index in Koalas DataFrame is valid, but I'm not sure whether we should crash in that case. E.g., `sdf.to_koalas()` doesn't have index, but should we explicitly `set_index` prior to `to_pandas()` in that case? This is the future work anyway. Let's file an issue to discuss it.
Thanks!

floscha commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,72 @@ def _unfold(key, kseries):
+        sdf = (series._kdf._sdf
+               .where(F.col(self._kdf._metadata.index_columns[0]) == row)
+               .select(column))
+        if sdf.count() < 1:
```
Good point! I've implemented your suggestion in 2dc8fa5.

floscha commented:
Thanks for your thorough review @ueshin!

I've addressed all of your comments and also added a warning note for matching a lot of rows with f14d136.

ueshin commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,80 @@ def _unfold(key, kseries):
+
+        if self._ks is None and (not isinstance(key, tuple) or len(key) != 2):
+            raise TypeError("Use DataFrame.at like .at[row_index, column_name]")
+        if self._ks is not None and len(key) != 1:
```
Need to check key is not str?

ueshin replied:
I mean what if `kdf.a.at['abc']`, for example?

floscha replied:
Good point! I've added the check and an additional unit test (that used to fail previously) with 
415e78a.

thunterdb commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,80 @@ def _unfold(key, kseries):
+    """
+    Access a single value for a row/column label pair.
+    Similar to ``loc``, in that both provide label-based lookups. Use ``at`` if you only need to
+    get a single value in a DataFrame or Series.
```
you should also mention the behavior when multiple values are matched (returning an array)

floscha replied:
I've added it to the docs with 0dc9e08.

thunterdb commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,80 @@ def _unfold(key, kseries):
+
+    Get value at specified row/column pair
+
+    >>> kdf.at[4, 'B']
```
can you also put an example with multiple values matching the same index, to show that it returns an array?

floscha replied:
Sure, done with 83a77d0.

ueshin commented on databricks/koalas/indexing.py:
```diff
@@ -82,6 +82,84 @@ def _unfold(key, kseries):
+
+    Get array if an index occurs multiple times
+
+    >>> kdf.to_pandas().at[5, 'B']
```
`kdf.at[5, 'B']` ?

floscha replied:
You're right. Fixed with ebea010.

ueshin commented on databricks/koalas/tests/test_indexing.py:
```diff
@@ -126,6 +126,45 @@ def pdf(self):
+        self.assertEqual(test_series.at['b'], 6)
+        self.assertEqual(test_series.at['b'], pdf.loc[3].at['b'])
+
+        #
```
forgot to add a comment?

floscha replied:
Correct. Guess I was kind of tired yesterday when committing this... Now fixed with ebea010 though.

ueshin commented:
Thanks! merging.
